tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+            offdiag[i] = recomb_coefficients.loc[i]
+        diag = np.hstack([np.zeros(1), -offdiag])
+        return (np.diag(diag) + np.diag(offdiag, k=1))[ion_number, :]
+
+    @staticmethod
+    def ion_matrix(ion_coefficients, atomic_number, ion_number):
+        """Constructs an ionization rate matrix from the ionization rates.
+
+        Parameters
+        ----------
+        ion_coefficients : DataFrame
+            Recombination coefficients.
+        atomic_number : int64
+            Current atomic number. Used for the dimension of a square matrix.
+        ion_number : int64
+            Current ion number. Used for returning the correct row.
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        offdiag = np.zeros(atomic_number)
+        index = ion_coefficients.index
+        for i in index:
+            offdiag[i] = ion_coefficients.loc[i]
+        diag = np.hstack([-offdiag, np.zeros(1)])
+        return (np.diag(diag) + np.diag(offdiag, k=-1))[ion_number, :]
+
+    @staticmethod
+    def prepare_charge_conservation_row(atomic_numbers):
+        """Prepares the last row of the rate_matrix. This row corresponds to the charge density equation."""
+        charge_conservation_row = []
+        for atomic_number in atomic_numbers:
+            charge_conservation_row.append(np.arange(0.0, atomic_number + 1))
+        charge_conservation_row = np.hstack([*charge_conservation_row, -1])
+        # TODO needs to be modified for use in nlte_excitation
+        return charge_conservation_row
+
+    @staticmethod
+    def prepare_ion_recomb_coefficients_nlte_ion(
+        gamma,
+        alpha_sp,
+        alpha_stim,
+        coll_ion_coeff,
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+    ):
+        """
+        Prepares the ionization and recombination coefficients by grouping them for ion numbers.
+        """
+        indexer = pd.Series(
+            np.arange(partition_function.shape[0]),
+            index=partition_function.index,
+        )
+        _ion2level_idx = indexer.loc[levels.droplevel("level_number")].values
+        partition_function_broadcast = partition_function.values[_ion2level_idx]
+        level_population_fraction = pd.DataFrame(
+            level_boltzmann_factor.values / partition_function_broadcast,
+            index=levels,
+        )
+        total_photo_ion_coefficients = (
+            (level_population_fraction.loc[gamma.index] * gamma)
+            .groupby(level=("atomic_number", "ion_number"))
+            .sum()
+        )
+        total_rad_recomb_coefficients = (
+            (alpha_sp + alpha_stim)
+            .groupby(level=["atomic_number", "ion_number"])
+            .sum()
+        )
+        total_coll_ion_coefficients = (
+            (
+                level_population_fraction.loc[coll_ion_coeff.index]
+                * coll_ion_coeff
+            )
+            .groupby(level=(0, 1))
```

Comment:
Does this dataframe not have named multi-index levels?

Reply:
Changed it